secure.go

```diff
@@ -235,53 +235,17 @@ func (secureProtocol *SecureProtocol) handleRegisterEx(packet nex.PacketInterfac
 		stationString, err := parametersStream.ReadString()
 
 		if err != nil {
-			go secureProtocol.RegisterExHandler(err, client, callID, stationUrls, NintendoLoginData{})
+			go secureProtocol.RegisterExHandler(err, client, callID, stationUrls, nex.NewDataHolder())
 			return
 		}
 
 		station := nex.NewStationURL(stationString)
 		stationUrls = append(stationUrls, station)
 	}
 
-	dataHolderType, err := parametersStream.ReadString()
+	dataHolder := parametersStream.ReadDataHolder()
 
-	if err != nil {
-		go secureProtocol.RegisterExHandler(err, client, callID, stationUrls, NintendoLoginData{})
-		return
-	}
-
-	if dataHolderType != "NintendoLoginData" {
-		err := errors.New("[SecureProtocol::RegisterEx] Data holder name does not match")
-		go secureProtocol.RegisterExHandler(err, client, callID, stationUrls, NintendoLoginData{})
-		return
-	}
-
-	if len(parametersStream.Bytes()[parametersStream.ByteOffset():]) < 8 {
-		err := errors.New("[SecureProtocol::RegisterEx] Data holder missing lengths")
-		go secureProtocol.RegisterExHandler(err, client, callID, stationUrls, NintendoLoginData{})
-		return
-	}
-
-	_ = parametersStream.ReadUInt32LE() // Length including next buffer length field
-	dataHolderInner, err := parametersStream.ReadBuffer()
-
-	if err != nil {
-		go secureProtocol.RegisterExHandler(err, client, callID, stationUrls, NintendoLoginData{})
-		return
-	}
-
-	dataHolderStream := nex.NewStreamIn(dataHolderInner, secureProtocol.server)
-
-	token, err := dataHolderStream.ReadString()
-
-	if err != nil {
-		go secureProtocol.RegisterExHandler(err, client, callID, stationUrls, NintendoLoginData{})
-		return
-	}
-
-	loginData := NintendoLoginData{Token: token}
-
-	go secureProtocol.RegisterExHandler(nil, client, callID, stationUrls, loginData)
+	go secureProtocol.RegisterExHandler(nil, client, callID, stationUrls, dataHolder)
```

Comment:
This dataholder can only contain a specific type. Even if the method is changed to pass the entire dataholder, the type name checking still needs to happen. Otherwise it will force every game server to do these checks resulting in lots of duplicate code across the network